activesupport/lib/active_support/duration.rb

```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
       def method_missing(method, *args, &block) #:nodoc:
         value.send(method, *args, &block)
       end
+
+    # Parses a string formatted according to ISO 8601 Duration into the hash .
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601DurationParser
+      attr_reader :parts
```

Comment:
It's a kind of this parser's API. That parts are being fed to duration in `ActiveSupport::Duration.parse!` method (see line 111)